File: core/src/main/java/xtdb/trie/LiveHashTrie.java
package xtdb.trie;

import org.apache.arrow.memory.util.ArrowBufPointer;
import org.apache.arrow.vector.ElementAddressableVector;

import java.util.Arrays;
import java.util.stream.IntStream;

public record LiveHashTrie(Node rootNode, ElementAddressableVector iidVec) implements HashTrie<LiveHashTrie.Node> {

    private static final int LOG_LIMIT = 64;
    private static final int PAGE_LIMIT = 1024;

    public sealed interface Node extends HashTrie.Node<Node> {
        Node add(LiveHashTrie trie, int idx);

        Node compactLogs(LiveHashTrie trie);
    }

    public static class Builder {
        private final ElementAddressableVector iidVec;
        private int logLimit = LOG_LIMIT;
        private int pageLimit = PAGE_LIMIT;

        private Builder(ElementAddressableVector iidVec) {
            this.iidVec = iidVec;
        }

        public void setLogLimit(int logLimit) {
            this.logLimit = logLimit;
        }

        public void setPageLimit(int pageLimit) {
            this.pageLimit = pageLimit;
        }

        public LiveHashTrie build() {
            return new LiveHashTrie(new Leaf(logLimit, pageLimit), iidVec);
        }
    }

    public static Builder builder(ElementAddressableVector iidVec) {
        return new Builder(iidVec);
    }

    @SuppressWarnings("unused")
    public static LiveHashTrie emptyTrie(ElementAddressableVector iidVec) {
        return new LiveHashTrie(new Leaf(LOG_LIMIT, PAGE_LIMIT), iidVec);
    }

    public LiveHashTrie add(int idx) {
        return new LiveHashTrie(rootNode.add(this, idx), iidVec);
    }

    public LiveHashTrie compactLogs() {
        return new LiveHashTrie(rootNode.compactLogs(this), iidVec);
    }

    private static final ThreadLocal<ArrowBufPointer> BUCKET_BUF_PTR = ThreadLocal.withInitial(ArrowBufPointer::new);

    private int bucketFor(int idx, int level) {
        return HashTrie.bucketFor(iidVec.getDataPointer(idx, BUCKET_BUF_PTR.get()), level);
    }

    private static final ThreadLocal<ArrowBufPointer> LEFT_BUF_PTR = ThreadLocal.withInitial(ArrowBufPointer::new);
    private static final ThreadLocal<ArrowBufPointer> RIGHT_BUF_PTR = ThreadLocal.withInitial(ArrowBufPointer::new);

    private int compare(int leftIdx, int rightIdx) {
        int cmp = iidVec.getDataPointer(leftIdx, LEFT_BUF_PTR.get()).compareTo(iidVec.getDataPointer(rightIdx, RIGHT_BUF_PTR.get()));
        if (cmp != 0) return cmp;

        // sort by idx desc
        return Integer.compare(rightIdx, leftIdx);
    }

    private static byte[] conjPath(byte[] path, byte idx) {
        int currentPathLength = path.length;
        var childPath = new byte[currentPathLength + 1];
        System.arraycopy(path, 0, childPath, 0, currentPathLength);
        childPath[currentPathLength] = idx;
        return childPath;
    }

    public record Branch(int logLimit, int pageLimit, byte[] path, Node[] children) implements Node {

        @Override
        public Node add(LiveHashTrie trie, int idx) {
            var bucket = trie.bucketFor(idx, path.length);

            var newChildren = IntStream.range(0, children.length)
                    .mapToObj(childIdx -> {
                        var child = children[childIdx];
                        if (bucket == childIdx) {
                            if (child == null) {
                                child = new Leaf(logLimit, pageLimit, conjPath(path, (byte) childIdx));
                            }
                            child = child.add(trie, idx);
                        }
                        return child;
                    }).toArray(Node[]::new);

            return new Branch(logLimit, pageLimit, path, newChildren);
        }

        @Override
        public Node compactLogs(LiveHashTrie trie) {
            var children =
                    Arrays.stream(this.children)
                            .map(child -> child == null ? null : child.compactLogs(trie))
                            .toArray(Node[]::new);

            return new Branch(logLimit, pageLimit, path, children);
        }
    }

    public record Leaf(int logLimit, int pageLimit, byte[] path, int[] data, int[] log, int logCount) implements Node {

        Leaf(int logLimit, int pageLimit) {
            this(logLimit, pageLimit, new byte[0]);
        }

        Leaf(int logLimit, int pageLimit, byte[] path) {
            this(logLimit, pageLimit, path, new int[0]);
        }

        private Leaf(int logLimit, int pageLimit, byte[] path, int[] data) {
            this(logLimit, pageLimit, path, data, new int[logLimit], 0);
        }

        @Override
        public Node[] children() {
            return null;
        }

        private int[] mergeSort(LiveHashTrie trie, int[] data, int[] log, int logCount) {
            int dataCount = data.length;

            var res = IntStream.builder();
            var dataIdx = 0;
            var logIdx = 0;

            while (true) {
                if (dataIdx == dataCount) {
                    IntStream.range(logIdx, logCount).forEach(idx -> res.add(log[idx]));
                    break;
                }

                if (logIdx == logCount) {
                    IntStream.range(dataIdx, dataCount).forEach(idx -> res.add(data[idx]));
                    break;
                }

                var dataKey = data[dataIdx];
                var logKey = log[logIdx];

                if (trie.compare(dataKey, logKey) <= 0) {
                    res.add(dataKey);
                    dataIdx++;
                } else {
                    res.add(logKey);
                    logIdx++;
                }
            }

            return res.build().toArray();
        }

        private int[] sortLog(LiveHashTrie trie, int[] log, int logCount) {
            // this is a little convoluted, but AFAICT this is the only way to guarantee a 'stable' sort,
            // (`Stream.sorted()` doesn't guarantee it), which is required for the log (to preserve insertion order)
            var boxedArray = Arrays.stream(log).limit(logCount).boxed().toArray(Integer[]::new);
            Arrays.sort(boxedArray, trie::compare);
            return Arrays.stream(boxedArray).mapToInt(i -> i).toArray();
        }

        private int[][] idxBuckets(LiveHashTrie trie, int[] idxs, byte[] path) {
            var entryGroups = new IntStream.Builder[LEVEL_WIDTH];
            for (int i : idxs) {
                int groupIdx = trie.bucketFor(i, path.length);
                var group = entryGroups[groupIdx];
                if (group == null) {
                    entryGroups[groupIdx] = group = IntStream.builder();
                }

                group.add(i);
            }

            return Arrays.stream(entryGroups).map(b -> b == null ? null : b.build().toArray()).toArray(int[][]::new);
        }

        @Override
        public Node compactLogs(LiveHashTrie trie) {
            if (logCount == 0) return this;

            var data = mergeSort(trie, this.data, sortLog(trie, log, logCount), logCount);
            var log = new int[this.logLimit];
            var logCount = 0;

            if (data.length > this.pageLimit) {
                var childBuckets = idxBuckets(trie, data, path);

                var childNodes = IntStream.range(0, childBuckets.length)
                        .mapToObj(childIdx -> {
                            var childBucket = childBuckets[childIdx];
                            return childBucket == null ? null : new Leaf(logLimit, pageLimit, conjPath(path, (byte) childIdx), childBucket);
                        }).toArray(Node[]::new);

                return new Branch(logLimit, pageLimit, path, childNodes);
            } else {
                return new Leaf(logLimit, pageLimit, path, data, log, logCount);
            }
        }

        @Override
        public Node add(LiveHashTrie trie, int newIdx) {
            var data = this.data;
            var log = this.log;
            var logCount = this.logCount;
            log[logCount++] = newIdx;
            var newLeaf = new Leaf(logLimit, pageLimit, path, data, log, logCount);

            return logCount == this.logLimit ? newLeaf.compactLogs(trie) : newLeaf;
        }
    }
}

File: core/src/main/kotlin/xtdb/trie/LiveHashTrie.kt
package xtdb.trie

import com.carrotsearch.hppc.IntArrayList
import org.apache.arrow.memory.util.ArrowBufPointer
import xtdb.trie.HashTrie.Companion.LEVEL_WIDTH
import xtdb.trie.HashTrie.Companion.bucketFor
import xtdb.vector.IVectorReader

private const val LOG_LIMIT = 64
private const val PAGE_LIMIT = 1024
private const val MAX_LEVEL = 64

data class LiveHashTrie(override val rootNode: Node, val iidReader: IVectorReader) : HashTrie<LiveHashTrie.Node> {
    interface Node : HashTrie.Node<Node> {
        fun add(trie: LiveHashTrie, newIdx: Int): Node

        fun compactLogs(trie: LiveHashTrie): Node
    }

    @Suppress("unused")
    class Builder(private val iidReader: IVectorReader) {
        private var logLimit = LOG_LIMIT
        private var pageLimit = PAGE_LIMIT
        private var rootPath = ByteArray(0)

        fun setLogLimit(logLimit: Int) = this.apply { this.logLimit = logLimit }
        fun setPageLimit(pageLimit: Int) = this.apply { this.pageLimit = pageLimit }
        fun setRootPath(path: ByteArray) = this.apply { this.rootPath = path }
        fun build(): LiveHashTrie = LiveHashTrie(Leaf(logLimit, pageLimit, rootPath), iidReader)
    }

    fun add(idx: Int) = copy(rootNode = rootNode.add(this, idx))

    @Suppress("unused")
    fun withIidReader(iidReader: IVectorReader) = copy(iidReader = iidReader)

    fun compactLogs() = copy(rootNode = rootNode.compactLogs(this))

    private fun bucketFor(idx: Int, level: Int): Int =
        bucketFor(iidReader.getPointer(idx, BUCKET_BUF_PTR.get()), level).toInt()

    private fun compare(leftIdx: Int, rightIdx: Int, leftPtr: ArrowBufPointer, rightPtr: ArrowBufPointer): Int {
        val cmp =
            iidReader.getPointer(leftIdx, leftPtr)
                .compareTo(iidReader.getPointer(rightIdx, rightPtr))

        return if (cmp != 0) cmp else rightIdx compareTo leftIdx
    }

    class Branch(
        private val logLimit: Int,
        private val pageLimit: Int,
        override val path: ByteArray,
        override val iidChildren: Array<Node?>,
    ) : Node {
        override val recencies = null
        override fun recencyNode(idx: Int) = throw UnsupportedOperationException()

        override fun add(trie: LiveHashTrie, newIdx: Int): Node {
            val bucket = trie.bucketFor(newIdx, path.size)

            val newChildren = iidChildren.indices
                .map { childIdx ->
                    var child = iidChildren[childIdx]
                    if (bucket == childIdx) {
                        child = child ?: Leaf(logLimit, pageLimit, conjPath(path, childIdx.toByte()))
                        child = child.add(trie, newIdx)
                    }
                    child
                }.toTypedArray()

            return Branch(logLimit, pageLimit, path, newChildren)
        }

        override fun compactLogs(trie: LiveHashTrie) =
            Branch(logLimit, pageLimit, path, iidChildren.map { child -> child?.compactLogs(trie) }.toTypedArray())
    }

    class Leaf(
        private val logLimit: Int,
        private val pageLimit: Int,
        override val path: ByteArray,
        val data: IntArray = IntArray(0),
        val log: IntArray = IntArray(logLimit),
        private val logCount: Int = 0,
        private var sortedData: IntArray? = null
    ) : Node {

        override val iidChildren = null

        override val recencies = null
        override fun recencyNode(idx: Int) = throw UnsupportedOperationException()

        fun mergeSort(trie: LiveHashTrie): IntArray {
            if (log.isEmpty()) return data
            return sortedData ?: mergeSort(trie, data, sortLog(trie, log, logCount), logCount)
        }

        private fun mergeSort(trie: LiveHashTrie, data: IntArray, log: IntArray, logCount: Int): IntArray {
            val leftPtr = ArrowBufPointer()
            val logPtr = ArrowBufPointer()
            val dataCount = data.size

            val res = IntArrayList(data.size + logCount)
            var dataIdx = 0
            var logIdx = 0

            while (true) {
                if (dataIdx == dataCount) {
                    for (idx in logIdx..<logCount) res.add(log[idx])
                    break
                }

                if (logIdx == logCount) {
                    for (idx in dataIdx..<dataCount) res.add(data[idx])
                    break
                }

                val dataKey = data[dataIdx]
                val logKey = log[logIdx]

                if (trie.compare(dataKey, logKey, leftPtr, logPtr) < 0) {
                    res.add(dataKey)
                    dataIdx++
                } else {
                    res.add(logKey)
                    logIdx++
                }
            }

            return res.toArray().also { sortedData = it }
        }

        private fun sortLog(trie: LiveHashTrie, log: IntArray, logCount: Int): IntArray {
            val leftPtr = ArrowBufPointer()
            val rightPtr = ArrowBufPointer()
            return log.take(logCount).sortedWith { leftKey: Int, rightKey: Int ->
                trie.compare(leftKey, rightKey, leftPtr, rightPtr)
            }.toIntArray()
        }


        private fun idxBuckets(trie: LiveHashTrie, idxs: IntArray, path: ByteArray): Array<IntArray?> {
            val entryGroups = arrayOfNulls<IntArrayList>(LEVEL_WIDTH)

            for (i in idxs) {
                val groupIdx = trie.bucketFor(i, path.size)
                val group = entryGroups[groupIdx] ?: IntArrayList().also { entryGroups[groupIdx] = it }
                group.add(i)
            }

            return entryGroups.map { b -> b?.toArray() }.toTypedArray()
        }

        override fun compactLogs(trie: LiveHashTrie): Node {
            if (logCount == 0) return this

            val data = if (sortedData != null) sortedData as IntArray else mergeSort(trie, data, sortLog(trie, log, logCount), logCount)
            val log = IntArray(logLimit)
            val logCount = 0

            return if (data.size > pageLimit && path.size < MAX_LEVEL) {
                val childBuckets = idxBuckets(trie, data, path)

                val childNodes = childBuckets
                    .mapIndexed<IntArray?, Node?> { childIdx, childBucket ->
                        if (childBucket == null) null
                        else
                            Leaf(logLimit, pageLimit, conjPath(path, childIdx.toByte()), childBucket)
                    }.toTypedArray()

                Branch(logLimit, pageLimit, path, childNodes)

            } else Leaf(logLimit, pageLimit, path, data, log, logCount)
        }

        override fun add(trie: LiveHashTrie, newIdx: Int): Node {
            var logCount = logCount
            log[logCount++] = newIdx
            val newLeaf = Leaf(logLimit, pageLimit, path, data, log, logCount)

            return if (logCount == logLimit) newLeaf.compactLogs(trie) else newLeaf
        }
    }

    companion object {
        @JvmStatic
        fun builder(iidReader: IVectorReader) = Builder(iidReader)

        @JvmStatic
        @Suppress("unused")
        fun emptyTrie(iidReader: IVectorReader) = builder(iidReader).build()

        private val BUCKET_BUF_PTR: ThreadLocal<ArrowBufPointer> = ThreadLocal.withInitial(::ArrowBufPointer)

        private fun conjPath(path: ByteArray, idx: Byte): ByteArray {
            val currentPathLength = path.size
            val childPath = ByteArray(currentPathLength + 1)
            System.arraycopy(path, 0, childPath, 0, currentPathLength)
            childPath[currentPathLength] = idx
            return childPath
        }
    }
}
